### DISCUSSION TITLE
Incremental primary constructor on record

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#7424] [Q&A]
```cs
record Person(string Name);
record Student(Person original, int Score) : Person(original);
var person = new Person("John");
var student = new Student(person, 5); // all good
// but...
Console.Write(nameof(Student.original)); // this is annoying
```
Is it possible without `original` property being synthesized?

**Comment by CyrusNajmabadi** [comment]
Sorry, it's not clear to me what is annoying here, it what you're asking for help on. Can you clarify?  Thanks!

**Reply by TahirAhmadov** [reply]
I'd like to declare a primary constructor on the "child" record which acts as an "upgrade" constructor, without passing in all of the properties of the "base" record, like below:
```cs
record Student(string Name, int Score) : Person(name);
```
However, when I just specify a parameter of the "base" record type, (like in the OP) the "child" record synthesizes a property of the "base" record type, (demonstrated by the `nameof()` line in the OP) which needlessly increases the size of the child record and maintains a reference, preventing that "base" record (which is no longer needed) from being GCed.

**Reply by HaloFour** [reply]
I don't think the synthesized copy constructor is intended to be used by external callers in that fashion.  A better approach would probably be to add a static factory method on the derived type that will destructor the record and then call the normal primary constructor.

I think making the copy constructors public, and what that entails, would be a conversation unto itself even without considering the additional request to support additional parameters.  But if you have a proposal as to what that might look like feel free to make a suggestion.  I have a feeling that it may come down to allowing the derived type to declare a secondary constructor that fits a specific pattern such that the compiler does not require it to call the primary constructor.  But I don't know how well something like that will work with primary constructors, and their various rules, being disconnected from positional records.

**Reply by TahirAhmadov** [reply]
Before providing any suggestions, I first wanted to find out if something like this is already possible elegantly. Off the top of my head, perhaps:
```cs
record Student(Person, int Score) : Person(...); 
```
Omit the parameter name, and pass it in using some new symbol, like `...`. 
Or:
```cs
record Student([DoNotCreate] Person original, int Score) : Person(original); 
record Student(noproperty Person original, int Score) : Person(original); 
record Student(base Person original, int Score) : Person(original); // re-use existing keyword
record Student(Person _, int Score) : Person(_); // this would be a breaking change
```

**Reply by CyrusNajmabadi** [reply]
Yes. What you want is possible. Just add a *normal* constructor to your derived record. Do not add the base type as a parameter to the *primary* constructor.

**Reply by HaloFour** [reply]
IMO such a copy constructor should never be the primary constructor as it does not define the schema of the type.  You'd either use inheritance, or composition, not both.

**Reply by HaloFour** [reply]
This already works:

```cs
record Person(string Name) { }

record Student(string Name, int Score) : Person(Name) {
    public Student(Person person, int score) : this(person.Name, score) { }
}
```

Obviously if the base record has a bunch of properties that would be more of a nuisance, but you could have a source generator that emits said secondary constructors.  The question is it worth having the compiler understand a different base constructor calling convention involving copy constructors specifically.

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi like this?
```cs
record Student2(string Name, int Score) : Person(Name)
{
  public Student2(Person original, int Score) : this(original.Name, Score)
  {
  }
}
```
That defeats the purpose - the whole idea here is to make it short and elegant, if I were to write all that out, I might as well just call the primary constructor directly:
```cs
record Student3(string Name, int Score) : Person(Name);
var person = new Person("John");
var student3 = new Student3(person.Name, 5);
```
The downside in both of the above is the need to repeat all the base record properties, which is what I'm trying to avoid. Always stay DRY.
PS. In my particular use case, the extra property and the GC complication are not bad enough and I'm going to use the approach in the OP for now. But I can imagine with a large number of instances it adds a real cost.

**Reply by CyrusNajmabadi** [reply]
> IMO such a copy constructor should never be the primary constructor as it does not define the schema of the type. You'd either use inheritance, or composition, not both.

Right. That's what I was saying :-)

**Reply by CyrusNajmabadi** [reply]
> That defeats the purpose - the whole idea here is to make it short and elegant, 


You're mixing composition and inheritance, which is pretty strange to begin with. If you want one, use one. If you want the other, just use the other. If you do really really really want to mix, well, having two constructors seems a perfectly fine way to bridge the two models to me.

**Reply by TahirAhmadov** [reply]
> IMO such a copy constructor should never be the primary constructor as it does not define the schema of the type. You'd either use inheritance, or composition, not both.

Yes it does; it says "take everything from the base record and add these new properties". The idea here is to use inheritance; the composition is the flaw I'm trying to get rid of.

**Reply by TahirAhmadov** [reply]
> > That defeats the purpose - the whole idea here is to make it short and elegant,
> 
> You're mixing composition and inheritance, which is pretty strange to begin with. If you want one, use one. If you want the other, just use the other. If you do really really really want to mix, well, having two constructors seems a perfectly fine way to bridge the two models to me.

I'm not using composition; the language imposes the composition aspect on me. Which is why something like this solves the problem:
```cs
record Student(base Person original, int Score) : Person(original); // re-use existing keyword
Console.Write(nameof(Student.original)); // error
```

**Reply by CyrusNajmabadi** [reply]
> The downside in both of the above is the need to repeat all the base record properties

But someone creating a student will need to pass them this information anyways, so how does this benefit anyone. 

If I'm generating a new student, I don't expect I'll do that by generating a person and generating a student *out of* that person.

A student "is a" person.  A student does not "have a" person.

**Reply by HaloFour** [reply]
> Always stay DRY.

Source generation tooling achieves this goal.

Again, the question is whether this approach should be _enshrined_ and _encouraged_ within the language.  I'd argue that it _probably_ doesn't, as it's a weird use of copy constructors outside of their intended purpose.

**Reply by TahirAhmadov** [reply]
> > The downside in both of the above is the need to repeat all the base record properties
> 
> But someone creating a student will need to pass them this information anyways, so how does this benefit anyone.
> 
> If I'm generating a new student, I don't expect I'll do that by generating a person and generating a student _out of_ that person.
> 
> A student "is a" person. A student does not "have a" person.

The person-student is just an example. The actual use case is taking an entity class from the persistence layer, and add some properties for presentation; imagine more of a `Person` -> `PersonModel`.

**Reply by CyrusNajmabadi** [reply]
> I'm not using composition

Yes you are. A record is *composed of* its constituent members. That's what a record *is*. 

If you don't want composition, then either don't use a record, or don't use the *primary constructor*. 

A simple secondary constructor solves the problem elegantly.  There is no composition change then, but you still expose this (Rather strange tbh) construction pattern to your consumers.

**Reply by TahirAhmadov** [reply]
> Source generation tooling achieves this goal.

How so? I still need to specify the child record with all the base record props, AND add an attribute and a source generator to create that secondary constructor. That's even more work. :)

> 
> Again, the question is whether this approach should be _enshrined_ and _encouraged_ within the language. I'd argue that it _probably_ doesn't, as it's a weird use of copy constructors outside of their intended purpose.

Define "weird" :)

**Reply by CyrusNajmabadi** [reply]
> and add some properties for presentation; 

So you want to model this (no pun intended) as an inheritance relationship or a composition one?  Personally, it sounds more like composition to me. But either could work.  But trying to make it both seems, ultimately, highly confusing.

**Reply by CyrusNajmabadi** [reply]
> Define "weird" :)

Something both being "is a" and "has a". Such types are not things normally seen.  You're well outside the norm here.

**Reply by CyrusNajmabadi** [reply]
> Which is why something like this solves the problem

The solution provided in the language today *also* solves the problem (quite elegantly imo). You get inheritance, and your simple way of creating new derived instances using values of the base type.

**Reply by TahirAhmadov** [reply]
> > I'm not using composition
> 
> Yes you are. 

Using the current syntax, yes that's the intention. Which is why I was hoping we can improve it to explicitly opt out of composition.

> A simple secondary constructor solves the problem elegantly. 

Not really - the objective is to avoid re-typing all the base record props, which secondary ctor only makes worse - now I have to list the props twice. (Instead of just listing them at the call site.)

> you still expose this (Rather strange tbh) construction pattern to your consumers.

I use "upgrade" ctors all the time. Without it, I need to declare these big model classes for presentation and copy most of the props from the underlying persistence layer class, which is extremely "wet". :)

**Reply by TahirAhmadov** [reply]
> So you want to model this (no pun intended) as an inheritance relationship or a composition one?

Inheritance only. Which is why I'd like to have a way to exclude a primary ctor parameter from generating a prop.

**Reply by TahirAhmadov** [reply]
Granted, I could use composition; then I would do:
```cs
record PersonModel(Person Person, string SomeDisplayProp);
```
However, then when I bind this to the UI or JSON serialization or whatever, I need to prefix everything with `Person.`, like `Person.Name`, instead of just `Name`, which is also "wet" and silly.

**Reply by CyrusNajmabadi** [reply]
> Using the current syntax, yes that's the intention. Which is why I was hoping we can improve it to explicitly opt out of composition.

You can. The approaches for doing this have already been presented. :-)

**Reply by TahirAhmadov** [reply]
> You can. The approaches for doing this have already been presented. :-)

I can't _elegantly,_ though. That's the problem. Those approaches involve repeating the base props for no reason - because I know I'm using the child record _only_ to add a few things to an existing base record.
Before records, I've been declaring classes, but there, it's even more boilerplate - create all the props, assign them in the ctor, the whole nine yards.

**Reply by CyrusNajmabadi** [reply]
> Not really - the objective is to avoid re-typing all the base record props, which secondary ctor only makes worse - now I have to list the props twice. (Instead of just listing them at the call site.)

You need to list them because your actual consumer needs to know how to create instances of your derived type.  

If I have student, I would not expect to have to instantiate a person first to make one. 

Same with your person/personmodel approach.  Why would I make a person instance just to throw it away to create the personmodel?

**Reply by TahirAhmadov** [reply]
> Same with your person/personmodel approach. Why would I make a person instance just to throw it away to create the personmodel?

The `Person` is returned from the ORM or API call; then I "upgrade" it to `PersonModel` for presentation.

**Reply by CyrusNajmabadi** [reply]
> I can't elegantly, though. 

It seems elegant to me. Note: it's not clear to me why you are even using positional parameters here in the first place. You are saying you care about positions, but then don't want people to use them. 

Why not just use normal property records here?

**Reply by TahirAhmadov** [reply]
> Why not just use normal property records here?

Too much typing :)

**Reply by CyrusNajmabadi** [reply]
> Those approaches involve repeating the base props for no reason 

It's not "no reason". You choose positional construction. So there needs to be a way to indicate how to positionally construct. 

But it sounds like you don't even care about positional construction, in which case just normal init properties and a special constructor for copying values over seems a better way to model this.

**Reply by CyrusNajmabadi** [reply]
> Too much typing :)

How so?

**Comment by TahirAhmadov** [comment]
To clarify what I mean, what I used to do before records:
```cs
// this is defined somewhere outside the UI
class Person
{
  public string Name { get; set; } // etc.
}
class PersonsLogic
{
  public Person GetPerson(int id) { ... }
}
// this is defined in the UI (like the web project)
class PersonModel : Person
{
  public PersonModel(Person source, string someDisplayProp)
  {
    AutoMapper.Copy(source, this);
    this.SomeDisplayProp = someDisplayProp;
  }
  public string SomeDisplayProp { get; }
}
// then: 
var person = this._businessLogic.GetPerson(123);
var personModel = new PersonModel(person, this.GetDisplayProp(person));
return this.View(personModel); // or whatever
```
It would be cool to replace the above with:
```cs
// this is defined somewhere outside the UI
record Person(string Name /* etc. */);
class PersonsLogic
{
  public Person GetPerson(int id) { ... }
}
// this is defined in the UI (like the web project)
record PersonModel(base Person source, string SomeDisplayProp) : Person(source);
// then: 
var person = this._businessLogic.GetPerson(123);
var personModel = new PersonModel(person, this.GetDisplayProp(person));
return this.View(personModel); // or whatever
```

**Reply by CyrusNajmabadi** [reply]
> It would be cool to replace the above with:

Why not just switch to get/init properties and use a record?

**Reply by spydacarnage** [reply]
I have achieved this exact model using a source generator - it's not open source, however, but it is very possible, and nicely neat to use.
The only different is that instead of a constructor, it creates an extension method on the source type with all the extra properties added as additional parameters, so you would use it as 
```cs
var person = this._businessLogic.GetPerson(123);
var personModel = person.ConvertToPersonModel(this.GetDisplayProp(person));
```

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi 

> Why not just switch to get/init properties and use a record?

Can you show an example of what you mean?

**Reply by TahirAhmadov** [reply]
@spydacarnage so it generates the extension method - basing on what? How do you define the "child" record?

**Reply by CyrusNajmabadi** [reply]
```c#
record Person
{
    public required string Name { get; init; }
};

record Student : Person
{
    public required int Score { get; init; }

    public static Student FromPerson(Person person, int score)
        => new()
        {
            Name = person.Name,
            Score = score,
        };
}

static void Main(string[] args)
{
    // Just create it directly.
    var student1 = new Student { Name = "foo", Score = 1 };

    // Create from an existing ORM object
    Person person = GetPerson();
    var student2 = Student.FromPerson(person, score: 1);
}
```

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi respectfully, that's even worse :))) I lose the primary ctor, I have to list all the properties, with all the boilerplate, and I have to create a static method on top of all that.
```cs
record Person(string Name /* etc. */);
record PersonModel(base Person source, string SomeDisplayProp) : Person(source);
```
This is what I'm talking about - 2 lines, voila.

**Reply by CyrusNajmabadi** [reply]
> This is what I'm talking about - 2 lines, voila.

Mixing composition and inheritance is a non-goal.  If you want htat, there are options to get it.  But we're not going to optimize with specific syntax just to save on the LOC here :)

**Reply by TahirAhmadov** [reply]
Again - by using the `base` keyword (or whatever syntax), we're saying that _**THIS IS NOT COMPOSITION**_. We're literally talking in circles. You say it's composition - and I have to repeat myself for the Nth time that this is exactly what I'm looking for - _not_ to compose. :)))

**Reply by HaloFour** [reply]
The primary constructor is supposed to define the schema of the record, and have symmetry with deconstruction.  That's not compatible with what you're proposing, which is to somehow "splat" the schema.

```cs
var student = new Student(person, 123);

if (student is Student(var o, var score)) {
    // what the heck is o?
}
```

**Reply by TahirAhmadov** [reply]
> // what the heck is o?

If it's `record Student(base Person original, int Score) : Person(original);`, then `o` is `Name` from `Person`.

**Reply by HaloFour** [reply]
Yeah, I really don't like that idea.  :)

**Reply by CyrusNajmabadi** [reply]
Yeah... That's really not good. Especially when Person is actually composed of 5 pieces of data.

**Reply by TahirAhmadov** [reply]
Both of you seem to think that the objective here is to deconstruct this record. That is not at all part of this use case.

**Reply by HaloFour** [reply]
> Both of you seem to think that the objective here is to deconstruct this record. That is not at all part of this use case.

Which is all the more reason to not use positional records here.  Positional deconstruction is one of their use cases.

**Reply by TahirAhmadov** [reply]
It's one of their use cases, but not a mandated one, is it? Or do you mean that by declaring a record but then not deconstructing it, i'm committing a sin of some sort? The gods of deconstruction will put me in high-ping, 100% CPU and out of memory hell? :)

**Reply by HaloFour** [reply]
> It's one of their use cases, but not a mandated one, is it? 

Yes, being able to deconstruct a positional record having symmetry with construction is one of their primary purposes and opinions.  Otherwise, I doubt positional records would have been added to the language at all.

**Reply by TahirAhmadov** [reply]
OK in that case, I'd like to confess, my father. I use records more and more, and don't have a single LoC where I deconstruct them. I beg lord for his mercy. :)

**Reply by HaloFour** [reply]
That is your choice, but that is the opinion of the design of records, and why you're going to find friction attempting to adapt positional records to this use case.

**Comment by spydacarnage** [comment]
> @spydacarnage so it generates the extension method - basing on what? How do you define the "child" record?

The `PersonModel` class is tagged with an attribute
```cs
[ConvertFrom<Person>]
public record PersonModel( ... )
```

The source generator then examines the two types, and fills out the method with simple assignments where the they have identical properties.  Extra properties in the destination are added to the method's parameter list.

**Reply by TahirAhmadov** [reply]
What is inside the `PersonModel` primary ctor?

**Reply by spydacarnage** [reply]
My version uses init props, like Cyrus' (but with the converter method source-generated).

Your Insistence on using PCs for this hampers your choices a little, but it might still be achievable - the SG would just need to do more work, and you'd have to carefully plan your types.

Or just switch to the method more suited to your requirements instead of trying to shoehorn the one that isn't...

**Reply by TahirAhmadov** [reply]
I see what you mean. Really though, in that case there is not much a record does for me that a regular class cannot, and the props have boilerplate of their own which PC parameters don't... What you have is better than nothing, but I think I prefer my approach even with the extraneous `Person original` property present on the child `PersonModel` record.
